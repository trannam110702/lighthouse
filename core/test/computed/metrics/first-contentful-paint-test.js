/**
 * @license
 * Copyright 2018 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from 'assert/strict';

import {FirstContentfulPaint} from '../../../computed/metrics/first-contentful-paint.js'; // eslint-disable-line max-len
import {getURLArtifactFromDevtoolsLog, readJson} from '../../test-utils.js';

const trace = readJson('../../fixtures/artifacts/progressive-app/trace.json', import.meta);
const devtoolsLog = readJson('../../fixtures/artifacts/progressive-app/devtoolslog.json', import.meta);

const URL = getURLArtifactFromDevtoolsLog(devtoolsLog);

describe('Metrics: FCP', () => {
  const gatherContext = {gatherMode: 'navigation'};

  it('should compute a simulated value', async () => {
    const settings = {throttlingMethod: 'simulate'};
    const context = {settings, computedCache: new Map()};
    const result = await FirstContentfulPaint.request(
      {trace, devtoolsLog, gatherContext, settings, URL},
      context);

    expect({
      timing: Math.round(result.timing),
      optimistic: Math.round(result.optimisticEstimate.timeInMs),
      pessimistic: Math.round(result.pessimisticEstimate.timeInMs),
      optimisticNodeTimings: result.optimisticEstimate.nodeTimings.size,
      pessimisticNodeTimings: result.pessimisticEstimate.nodeTimings.size,
    }).toMatchSnapshot();
    assert.ok(result.optimisticGraph, 'should have created optimistic graph');
    assert.ok(result.pessimisticGraph, 'should have created pessimistic graph');
  });

  it('should compute an observed value (desktop)', async () => {
    const settings = {throttlingMethod: 'provided', formFactor: 'desktop'};
    const context = {settings, computedCache: new Map()};
    const result = await FirstContentfulPaint.request({trace, devtoolsLog, gatherContext, settings},
      context);

    await expect(result).toMatchInlineSnapshot(`
Object {
  "timestamp": 376406173872,
  "timing": 192.308,
}
`);
  });

  it('should compute an observed value (mobile)', async () => {
    const settings = {throttlingMethod: 'provided', formFactor: 'mobile'};
    const context = {settings, computedCache: new Map()};
    const result = await FirstContentfulPaint.request(
      {gatherContext, trace, devtoolsLog, settings}, context);

    await expect(result).toMatchInlineSnapshot(`
Object {
  "timestamp": 376406173872,
  "timing": 192.308,
}
`);
  });
});
