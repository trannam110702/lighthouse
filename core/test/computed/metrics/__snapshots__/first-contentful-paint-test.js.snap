// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics: FCP should compute a simulated value 1`] = `
Object {
  "optimistic": 1107,
  "optimisticNodeTimings": 4,
  "pessimistic": 1107,
  "pessimisticNodeTimings": 4,
  "timing": 1107,
}
`;
